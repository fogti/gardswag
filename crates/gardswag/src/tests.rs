use super::*;

fn dflsubscr() -> impl tracing::subscriber::Subscriber {
    tracing_subscriber::fmt::Subscriber::builder()
        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
        .finish()
}

macro_rules! assert_interp {
    ($x:expr) => {{
        let result = crossbeam_utils::thread::scope(|s| main_interp(s, $x))
            .expect("unable to join threads");
        insta::assert_debug_snapshot!(result);
    }}
}

#[test]
fn chk_hello() {
    insta::assert_yaml_snapshot!(main_check(r#"std.stdio.write("Hello world!\n");"#).unwrap());
}

#[test]
fn chk_fibo() {
    tracing::subscriber::with_default(dflsubscr(), || {
        insta::assert_yaml_snapshot!(main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib
            "#
        )
        .unwrap());
    });
}

#[test]
fn run_fibo0() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 0
            "#,
        )
        .unwrap();
        assert_interp!(&x.0);
    });
}

#[test]
fn run_fibo1() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 1
            "#,
        )
        .unwrap();
        assert_interp!(&x.0);
    });
}

#[test]
fn run_fibo() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 5
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn chk_implicit_restr() {
    tracing::subscriber::with_default(dflsubscr(), || {
        insta::assert_yaml_snapshot!(main_check(
            r#"
                \x
                let id = \y y;
                .{
                  id;
                  x;
                  y = "{x}";
                }
            "#
        )
        .unwrap());
    });
}

#[test]
fn run_id() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let id = \x x;
                id 1
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn run_call_blti() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                std.plus 1 1
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn run_fix() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec f = \a if (std.eq a 0) { 0 } { f 0 };
                f 1
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn run_update() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                .{
                  a = "what";
                  b = 1;
                  c = .{};
                } // .{
                  b = "no";
                  c = 50;
                }
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn error_int_update() {
    tracing::subscriber::with_default(dflsubscr(), || {
        insta::assert_yaml_snapshot!(main_check("0//0").map_err(|e| e.to_string()));
    });
}

#[test]
fn run_ctrl_match() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check("match .this_is_a_variant 1 | .this_is_a_variant x => std.plus x 1")
            .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

proptest::proptest! {
    #![proptest_config(proptest::test_runner::Config::with_cases(8192))]

    #[test]
    fn doesnt_crash(s in "[ -~]+") {
        if let Ok(x) = main_check(&s) {
            let _ = crossbeam_utils::thread::scope(|s| main_interp(s, &x.0))
                .expect("unable to join threads");
        }
    }
}
