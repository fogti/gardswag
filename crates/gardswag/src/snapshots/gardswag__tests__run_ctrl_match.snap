---
source: crates/gardswag/src/tests.rs
expression: x
---
- stmts: []
  term:
    offset: 0
    inner:
      Match:
        inp:
          offset: 6
          inner:
            Call:
              prim:
                offset: 6
                inner:
                  Tagger:
                    key: this_is_a_variant
                extra:
                  ty:
                    Arrow1:
                      - Literal: Int
                      - TaggedUnion:
                          this_is_a_variant:
                            Literal: Int
                  ident_multi: 0
              arg:
                offset: 25
                inner:
                  Integer: 1
                extra:
                  ty:
                    Literal: Int
                  ident_multi: 0
          extra:
            ty:
              TaggedUnion:
                this_is_a_variant:
                  Literal: Int
            ident_multi: 0
        cases:
          - pat:
              Tagged:
                key:
                  offset: 30
                  inner: this_is_a_variant
                  extra: ~
                value:
                  Identifier:
                    offset: 48
                    inner: x
                    extra:
                      ty:
                        Literal: Int
                      ident_multi: 1
            body:
              offset: 56
              inner:
                Call:
                  prim:
                    offset: 56
                    inner:
                      Call:
                        prim:
                          offset: 56
                          inner:
                            Dot:
                              prim:
                                offset: 53
                                inner:
                                  Identifier: std
                                extra:
                                  ty:
                                    Record:
                                      chan_recv:
                                        Arrow1:
                                          - ChanRecv:
                                              Var: 13
                                          - TaggedUnion:
                                              None:
                                                Literal: Unit
                                              Some:
                                                Var: 13
                                      chan_send:
                                        Arrow1:
                                          - Var: 12
                                          - Arrow1:
                                              - ChanSend:
                                                  Var: 12
                                              - Literal: Bool
                                      eq:
                                        Arrow1:
                                          - Var: 9
                                          - Arrow1:
                                              - Var: 10
                                              - Literal: Bool
                                      leq:
                                        Arrow1:
                                          - Literal: Int
                                          - Arrow1:
                                              - Literal: Int
                                              - Literal: Bool
                                      make_chan:
                                        Arrow0:
                                          - Literal: Unit
                                          - Record:
                                              recv:
                                                ChanRecv:
                                                  Var: 11
                                              send:
                                                ChanSend:
                                                  Var: 11
                                      minus:
                                        Arrow1:
                                          - Literal: Int
                                          - Arrow1:
                                              - Literal: Int
                                              - Literal: Int
                                      mult:
                                        Arrow1:
                                          - Literal: Int
                                          - Arrow1:
                                              - Literal: Int
                                              - Literal: Int
                                      not:
                                        Arrow1:
                                          - Literal: Bool
                                          - Literal: Bool
                                      plus:
                                        Arrow1:
                                          - Literal: Int
                                          - Arrow1:
                                              - Literal: Int
                                              - Literal: Int
                                      spawn_thread:
                                        Arrow1:
                                          - ArrowU:
                                              - Literal: Unit
                                              - Literal: Unit
                                          - Literal: Unit
                                      stdio:
                                        Record:
                                          write:
                                            Arrow1:
                                              - Literal: String
                                              - Literal: Unit
                                  ident_multi: 0
                              key:
                                offset: 57
                                inner: plus
                                extra: ~
                          extra:
                            ty:
                              Arrow1:
                                - Literal: Int
                                - Arrow1:
                                    - Literal: Int
                                    - Literal: Int
                            ident_multi: 0
                        arg:
                          offset: 62
                          inner:
                            Identifier: x
                          extra:
                            ty:
                              Literal: Int
                            ident_multi: 0
                    extra:
                      ty:
                        Arrow1:
                          - Literal: Int
                          - Literal: Int
                      ident_multi: 0
                  arg:
                    offset: 64
                    inner:
                      Integer: 1
                    extra:
                      ty:
                        Literal: Int
                      ident_multi: 0
              extra:
                ty:
                  Literal: Int
                ident_multi: 0
    extra:
      ty:
        Literal: Int
      ident_multi: 0
- forall: []
  forall_args: []
  ty:
    Literal: Int

